File: src/net/masterthought/dlanguage/module/DubInitForModuleStep.java
package net.masterthought.dlanguage.module;

import com.intellij.ide.util.projectWizard.ModuleBuilder;
import com.intellij.ide.util.projectWizard.ModuleWizardStep;
import com.intellij.ide.util.projectWizard.ProjectBuilder;
import com.intellij.ide.util.projectWizard.WizardContext;
import com.intellij.openapi.ui.ComboBox;
import com.intellij.openapi.util.Pair;
import com.intellij.projectImport.ProjectFormatPanel;

import javax.swing.*;
import java.awt.*;
import java.util.ArrayList;
import java.util.List;

public class DubInitForModuleStep extends ModuleWizardStep {

    private final JPanel myPanel;
    protected final WizardContext myWizardContext;
    private final ProjectFormatPanel myFormatPanel = new ProjectFormatPanel();
    private final ComboBox dubFormat;
    private final ComboBox dubType;
    private final JTextField dubParams;

    public DubInitForModuleStep(WizardContext wizardContext) {
        this.myWizardContext = wizardContext;
        this.myPanel = new JPanel(new GridBagLayout());
        this.myPanel.setBorder(BorderFactory.createEmptyBorder(10, 10, 10, 10));

        // Title label
        JLabel titletextLabel = new JLabel("Please choose the dub init options you prefer");
        this.myPanel.add(titletextLabel, new GridBagConstraints(0, -1, 1, 1, 1.0D, 0.0D, 18, 2, new Insets(8, 10, 8, 10), 0, 0));

        // add combo to choose --format sdl/json for init
        java.util.List<String> options = new ArrayList<>();
        options.add("sdl");
        options.add("json");
        String[] opts = options.toArray(new String[options.size()]);
        this.dubFormat = new ComboBox(opts);
        dubFormat.setSelectedItem("json");

        JLabel dubFormatLabel = new JLabel("Dub format");
        dubFormatLabel.setLabelFor(dubFormat);

        this.myPanel.add(dubFormatLabel, new GridBagConstraints(0, -1, 1, 1, 0.0D, 0.0D, 17, 0, new Insets(0, 0, 5, 4), 0, 0));
        this.myPanel.add(dubFormat, new GridBagConstraints(0, -1, 1, 1, 1.0D, 0.0D, 18, 2, new Insets(10, 0, 20, 0), 0, 0));

        // add combo to choose --type minimal, vibe.d, deimos
        java.util.List<String> typeOptions = new ArrayList<>();
        typeOptions.add("minimal");
        typeOptions.add("vibe.d");
        typeOptions.add("deimos");
        String[] typeOpts = typeOptions.toArray(new String[typeOptions.size()]);
        this.dubType = new ComboBox(typeOpts);
        dubType.setSelectedItem("minimal");

        JLabel dubTypeLabel = new JLabel("Dub project type");
        dubTypeLabel.setLabelFor(dubType);

        this.myPanel.add(dubTypeLabel, new GridBagConstraints(0, -1, 1, 1, 0.0D, 0.0D, 17, 0, new Insets(0, 0, 5, 4), 0, 0));
        this.myPanel.add(dubType, new GridBagConstraints(0, -1, 1, 1, 1.0D, 0.0D, 18, 2, new Insets(10, 0, 20, 0), 0, 0));

        // add text field to add params instead of using combo boxes as fallback
        this.dubParams = new JTextField();
        JLabel dubParamsLabel = new JLabel("Dub params override - supply space separated params here instead and it will use these e.g --format=json");
        dubParamsLabel.setLabelFor(dubParams);
        this.myPanel.add(dubParamsLabel, new GridBagConstraints(0, -1, 1, 1, 1.0D, 0.0D, 18, 2, new Insets(10, 0, 20, 0), 0, 0));
        this.myPanel.add(dubParams, new GridBagConstraints(0, -1, 1, 1, 1.0D, 0.0D, 18, 2, new Insets(10, 0, 20, 0), 0, 0));

    }

    public JComponent getComponent() {
        return this.myPanel;
    }

    public boolean isStepVisible() {
        return this.myWizardContext.getProject() == null;
    }

    public void updateDataModel() {
        ProjectBuilder moduleBuilder = this.myWizardContext.getProjectBuilder();
        if (moduleBuilder != null) {
            this.myWizardContext.setProjectBuilder(moduleBuilder);
            if (moduleBuilder instanceof ModuleBuilder) {
                ModuleBuilder builder = (ModuleBuilder) moduleBuilder;

                if (builder.getBuilderId() != null && builder.getBuilderId().equals("DLangDubApp")) {
                    DLanguageDubModuleBuilder dubBuilder = (DLanguageDubModuleBuilder) builder;

                    List<Pair<String, String>> optionsList = new ArrayList<>();
                    optionsList.add(Pair.create("dubFormat", this.dubFormat.getSelectedItem().toString()));
                    optionsList.add(Pair.create("dubType", this.dubType.getSelectedItem().toString()));
                    optionsList.add(Pair.create("dubParams", this.dubParams.getText()));

                    dubBuilder.setDubInitOptions(optionsList);
                }
            }
        }

        this.myFormatPanel.updateData(this.myWizardContext);
    }

    public Icon getIcon() {
        return this.myWizardContext.getStepIcon();
    }

}

File: dub/src/main/kotlin/io/github/intellij/dub/module/DubInitForModuleStep.kt
package io.github.intellij.dub.module

import com.intellij.ide.util.projectWizard.ModuleBuilder
import com.intellij.ide.util.projectWizard.ModuleWizardStep
import com.intellij.ide.util.projectWizard.WizardContext
import com.intellij.openapi.ui.ComboBox
import com.intellij.util.ui.JBUI
import io.github.intellij.dlanguage.DlangBundle.message
import io.github.intellij.dlanguage.module.DlangProjectFormatPanel
import java.awt.GridBagConstraints
import java.awt.GridBagLayout
import javax.swing.*

class DubInitForModuleStep(protected val myWizardContext: WizardContext) : ModuleWizardStep() {
    private val myPanel: JPanel
    private val myFormatPanel = DlangProjectFormatPanel()
    private val dubFormat: ComboBox<String>
    private val dubType: ComboBox<String>
    private val dubParams: JTextField

    init {
        myPanel = JPanel(GridBagLayout())
        myPanel.border = BorderFactory.createEmptyBorder(10, 10, 10, 10)

        // Title label
        val titletextLabel = JLabel(message("d.ui.dub.config.label.choosedubinitoptions"))
        myPanel.add(titletextLabel, GridBagConstraints(0, -1, 1, 1, 1.0, 0.0, 18, 2, JBUI.insets(8, 10, 8, 10), 0, 0))

        // add combo to choose --format sdl/json for init
        dubFormat = ComboBox(arrayOf("sdl", "json"))
        dubFormat.selectedItem = "json"
        val dubFormatLabel = JLabel(message("d.ui.dub.config.label.dubformat"))
        dubFormatLabel.labelFor = dubFormat
        myPanel.add(dubFormatLabel, GridBagConstraints(0, -1, 1, 1, 0.0, 0.0, 17, 0, JBUI.insets(0, 0, 5, 4), 0, 0))
        myPanel.add(dubFormat, GridBagConstraints(0, -1, 1, 1, 1.0, 0.0, 18, 2, JBUI.insets(10, 0, 20, 0), 0, 0))

        // add combo to choose --type minimal, vibe.d, deimos
        dubType = ComboBox(arrayOf("minimal", "vibe.d", "deimos"))
        dubType.selectedItem = "minimal"
        val dubTypeLabel = JLabel(message("d.ui.dub.config.label.dubprojecttype"))
        dubTypeLabel.labelFor = dubType
        myPanel.add(dubTypeLabel, GridBagConstraints(0, -1, 1, 1, 0.0, 0.0, 17, 0, JBUI.insets(0, 0, 5, 4), 0, 0))
        myPanel.add(dubType, GridBagConstraints(0, -1, 1, 1, 1.0, 0.0, 18, 2, JBUI.insets(10, 0, 20, 0), 0, 0))

        // add text field to add params instead of using combo boxes as fallback
        dubParams = JTextField()
        val dubParamsLabel = JLabel(message("d.ui.dub.config.label.dubparamsoverride"))
        dubParamsLabel.labelFor = dubParams
        myPanel.add(dubParamsLabel, GridBagConstraints(0, -1, 1, 1, 1.0, 0.0, 18, 2, JBUI.insets(10, 0, 20, 0), 0, 0))
        myPanel.add(dubParams, GridBagConstraints(0, -1, 1, 1, 1.0, 0.0, 18, 2, JBUI.insets(10, 0, 20, 0), 0, 0))
    }

    override fun getComponent(): JComponent {
        return myPanel
    }

    override fun isStepVisible(): Boolean {
        return myWizardContext.project == null
    }

    override fun updateDataModel() {
        val moduleBuilder = myWizardContext.projectBuilder
        if (moduleBuilder != null) {
            myWizardContext.projectBuilder = moduleBuilder
            if (moduleBuilder is ModuleBuilder) {
                val builder = moduleBuilder
                if ("DLangDubApp" == builder.builderId) {
                    val dubBuilder = builder as DlangDubModuleBuilder
                    val options: MutableMap<String, String> = HashMap(3)
                    options["dubFormat"] = dubFormat.selectedItem.toString()
                    options["dubType"] = dubType.selectedItem.toString()
                    options["dubParams"] = dubParams.text
                    dubBuilder.setDubOptions(options)
                }
            }
        }
        myFormatPanel.updateData(myWizardContext)
    }

    override fun getIcon(): Icon? {
        return myWizardContext.stepIcon
    }
}
